server/aesdsocket.c

```diff
@@ -197,67 +340,80 @@ int main(int argc, char *argv[])
 
-        // Open file for appending
-        int file_fd = open(FILE_PATH, O_WRONLY | O_CREAT | O_APPEND, 0644);
-        if (file_fd == -1) {
-            syslog(LOG_ERR, "Failed to open file");
+        // Allocate a new thread_info structure for this connection
+        struct thread_info *tinfo = malloc(sizeof(struct thread_info));
+        if (!tinfo) {
+            syslog(LOG_ERR, "Failed to allocate memory for thread info");
             close(client_fd);
             continue;
         }
-        
-        // Receive data and write to file
-        while ((bytes_received = recv(client_fd, buffer, BUFFER_SIZE, 0)) > 0) {
-            if (write(file_fd, buffer, bytes_received) == -1) {
-                syslog(LOG_ERR, "Failed to write to file");
-                break;
-            }
+        tinfo->client_fd = client_fd;
 
-            // Check for end of packet based on '\n' as the delimiter
-            if (buffer[bytes_received - 1] == '\n') {
-                break;
-            }
-        }
-        
-        // Close file after writing
-        close(file_fd);
+        // Add the thread info to the global list
+        pthread_mutex_lock(&list_mutex);
+        LIST_INSERT_HEAD(&thread_list, tinfo, entries);
+        pthread_mutex_unlock(&list_mutex);
 
-        // Reopen file for Reading Content
-        file_fd = open(FILE_PATH, O_RDONLY);
-        if (file_fd == -1) {
-            syslog(LOG_ERR, "Failed to open file for reading");
+         // Create a thread to handle the client
+         if (pthread_create(&tinfo->thread_id, NULL, handle_client, tinfo) != 0) {
+            syslog(LOG_ERR, "Failed to create thread");
+            pthread_mutex_lock(&list_mutex);
+            LIST_REMOVE(tinfo, entries);
+            pthread_mutex_unlock(&list_mutex);
             close(client_fd);
+            free(tinfo);
             continue;
-        }
-
-        // Read the file and send its content to the client
-        while ((bytes_received = read(file_fd, buffer, BUFFER_SIZE)) > 0) {
-            send(client_fd, buffer, bytes_received, 0);
-        }
+         }
+    }
 
-        // Close file after reading
-        close(file_fd);
+    if (server_fd != -1){
+        close(server_fd);
+    }
 
-        syslog(LOG_INFO, "Closed connection from %s", inet_ntoa(client_in->sin_addr));
+    /* Cancel and join the timer thread */
+    pthread_cancel(timer_thread_id);
+    pthread_join(timer_thread_id, NULL);
 
-        // Close Client Connection
-        close(client_fd);
+    /* Request exit from each thread by canceling them and then join */
+    pthread_mutex_lock(&list_mutex);
+    struct thread_info *entry;
+    for (entry = LIST_FIRST(&thread_list); entry != NULL; ) {
+        struct thread_info *next_entry = LIST_NEXT(entry, entries);
+        pthread_cancel(entry->thread_id);
+        pthread_mutex_unlock(&list_mutex);
+        pthread_join(entry->thread_id, NULL);
```

Comment:
It doesn't look like you really need a lock given that you only access the list from one thread.  However if you did you might have an issue here given that you are unlocking the list mutex while iterating over it.